lambda_functions/healthcheck/healthcheck.py

```diff
@@ -0,0 +1,126 @@
+
+    :param Exception: Exception message to be logged to the healthcheck lambda log group
+    :type Exception: string
+    """    
+    def __init__(self, message):
+        self.message = message
+        super().__init__(self.message)
+
+    def __str__(self):
+        return f'Custom Exception: {self.message}'
+
+def failing_metric(url, env, expected_string, status_code, failure_type):
+    """
+    failing_metric This function wll be triggered if our healthcheck fails.
+    It will log custom metric data to AWS Cloudwatch Metrics
+
+    :param url: Url of url to be checked
+    :type url: string
+    :param expected_string: String of text to be checked
+    :type expected_string: String
+    :param env: This is for us to distinguish the message between the dev, preprod and prod environments
+    :type env: string
+    :param status_code: The status code value we received from the healthcheck.
+    :type status_code: string
+    :param failure_type: The failure type helps to distinguish between a response and content failure for metric collection.
+    :type failure_type: string
+    """    
+    # Define metric data
+    cloudwatch = boto3.client('cloudwatch')
+    metric_data = cloudwatch.put_metric_data(
+        MetricData = [
+            {
+                'MetricName': f'{env}-healthcheck-failures',
+                'Dimensions': [
+                    {
+                    'Name': 'EndpointUrl',
+                    'Value': url,
+                    },
+                    {
+                    'Name': 'ExpectedString',
+                    'Value': expected_string,
+                    },
+                    {
+                    'Name': 'ResponseStatus',
+                    'Value': str(status_code),
+                    },
+                    {
+                    'Name': 'FailureType',
+                    'Value': failure_type,
+                    },
+                ],
+                'Unit': 'Count',
+                'Value': 1,
+            },
+        ],
+        Namespace = 'SML-Healthcheck.'
+    )
+
+    print("Metric Data: ", metric_data)
+
+def check_web_url_health(url, env, expected_string):
+    """
+    check_web_url_health is a function that calls the url from the event
+    and checks if the status code is 200 and the expected string exists on that url.
+
+    :param url: Url of url to be checked
+    :type url: string
+    :param expected_string: String of text to be checked
+    :type expected_string: String
+    :param env: This is for us to distinguish the message between the dev, preprod and prod environments
+    :type env: string
+    :param failure_type: The failure type helps to distinguish between a response and content failure for metric collection.
+    :type failure_type: string
+    :raises HealthCheckException: this custom exception will trigger if the response does not match our expected values.
+    :raises HealthCheckException: error
+    """
+    response = requests.get(url, timeout=5)
```

Comment:
Where has the try/catch block gone?  You need to catch any exceptions from the request package.

The aim is to handle exceptions that come from the requests package as well as raise exceptions when the response we receive doesn't match what we expect.

You want something like this (pseudo code):
```
try:
   response = requests.get(url)
   
   if ! 200:
     raise HealthCheckException(unexpected response code message)
     
   if ! expected_string:
     raise HealthCheckException(unexpected content message)
except requests.RequestException as e:
   raise HealthCheckException(f"Request failed: {e}")
except Exception as e:
   raise HealthCheckException(f"Unexpected error: {e}")
```

Reply:
done :)